### ISSUE TITLE
Colored route line sometimes vanishes sporadically

### ISSUE THREAD

**Issue by Norbert80** [osmandapp/Osmand#1338]
Colored route line sometimes vanishes sporadically at certain zoom level.
@vshcherb I remember that 2.0 pre releases had similar problems but more often appearing.

Found with OsmAndF~2.1.0#7464D.

![screenshot_2015-06-02-18-23-25](https://cloud.githubusercontent.com/assets/6921696/8117277/2e09536a-1087-11e5-8fed-52bd177dc705.png)

**Comment by razor74** [comment]
The same problem to me on OsmandF~2.1.0.#7740D

**Comment by vshcherb** [comment]
Is there way to reproduce without navigation (movement)? Is it constantly reproducible in 1 place (zoom)?

**Comment by razor74** [comment]
Without movement not dissapear. Not constantly. Sometimes i drive few 
kilometers without dissapearing. It dissapear when map refreshing from 
moving...

**Comment by vshcherb** [comment]
Unfortunately this bug is not reproducible, so there is no action planned yet. Once we've got more information by testing we will reopen and fix it.

**Comment by Norbert80** [comment]
@vshcherb another screen shot in italy (toscany).
![screenshot_2015-06-28-23-03-10](https://cloud.githubusercontent.com/assets/6921696/8528004/7420cf3a-2410-11e5-8478-95ffb5cb71f5.png)

**Comment by vshcherb** [comment]
This issue is closed, but doesn't mean it is not active. Last 2 weeks doing lots of testing, I saw that issue as well. It happened in the navigation and in few moments disappeared, I managed to take screenshot too, but that didn't help later. So the issue is still not reproducible, the latest news that it might be harder reproduce or even not possible, due to latest fixes around ui memory leaks (which are easy to reproduce and they were fixed, please check another issue with black screen).

**Comment by sonora** [comment]
From what I can see this issue may be unrelated to our memory-leak-upon-screen-rotation issue, because I 95% use my devices without screen rotation, and hence never experienced the map-goes-blank or memory leak issue, but I have seen this route-disappears occasionally. Also, route re-appears after 3 seconds or so, which would also be untypical.

The issue I does not seem to show up when you are viewing overview zooms or the whole route, so I would say there are certain conditions under which only the first route segment (maybe from your position to the next turn?) disappears before immediately being re-drawn.

**Comment by Norbert80** [comment]
I can confirm the comment from @sonora.

**Comment by vshcherb** [comment]
In my situations I saw it only on very high level zooms, like >= 17th zoom. I remember when we got this issue reproducible and it was due to Android rendering engine. It renders that line as a big texture which sometimes doesn't video memory. After that we started cutting the line as much as possible, only up to top of the screen. 
However the issue could be related to another fix was done, we reduced quality of the image from ARGB_8888 to RGB_565 (when transparency is not used), so instead of 18 MB per image we have 4 MB Image, so that could definitely help, as I see.

**Comment by razor74** [comment]
Today i encounter again the issue but after a long time driving with phone on portrait position... After i rotate the phone on landscape and re-rotate on portrait the route line dissapear repeatedly at almost regulate time for 4-5 secconds... I use the map with autozooom disabled. The bug encounter at lower zoom  ( even 50 m) or up - 100 m, 200 m  etc.

**Comment by sonora** [comment]
For what it is worth: I have tested OsmAnd for about 5000 km in recent weeks, and for this particular issue I can now say it seems to happen when the (from your current position) immediately next route segment is perfectly straight for a long stretch (like one or several kilometers), as it occurs when you are on a motorway through a flat area with no curves coming up for long enough. This is also visible on the screenshots above, note that the route segment is straight all the way beyond the screen edge.

Once you approach a turn in the route, or even just a curve of the road, the phenomenon disappears again. You can also make it disappear if you manually zoom the map out far enough, so that turns or curves become visible on the screen (any dependency on the map magnifier value untested, I use our defaults). Zooming back in makes the bug re-appear.

Contrary to previous guesses, the issue is independent of screen rotation settings, i.e. also happens in the static "North is up" mode.

So it seems some sort of code bug where we do not draw the first segment of the route line to start immediately at the "My Position" icon, or maybe not a all, if it is straight beyond the screen edge. Funny enough, as the screenshots show, the little black arrows remain always visible, only the blue transparent line disappears. So maybe we can find the bug by comparing with what code we draw the black arrows vs. the blue line,

**Comment by andrew-davie** [comment]
Yes, this is due to the clipping of the line - you can see the same effect when you get the endpoint of a line very close to the edge of screen - the next part of the line disappears too early. I rewrote the route code completely and came across this bug in the original code when I was doing that.

**Comment by Norbert80** [comment]
@andrew-davie do you have a fix available with your rewritten code you are talking about?

**Comment by andrew-davie** [comment]
@Norbert80 yes, I rewrote the code and it was working pretty well. But it sat unloved as a pull request for too long, so I removed it.  Still sitting in a branch somewhere.  I think the project was just too busy with the Live stuff to address my new stuff, and I thought best to simply remove the "pressure".  My changes are a complete rewrite not a "fix".  I posted a few videos - for example, https://www.youtube.com/watch?v=0OQ7uR3TtRU

**Comment by Norbert80** [comment]
@andrew-davie thanks, I have watched your video. But I think that in current OsmAnd builds the arrows are more exactly aligned with the route.

**Comment by andrew-davie** [comment]
@Norbert80 the subsampling of the track into evenly spaced segments for the purposes of arrow display (i.e, every 5 m) can cause an impression when there is a lot of GPX 'jiggle' that they are not perfectly aligned. But in the video shown I have a fairly course sampling (say, 10 m). That might be what you are seeing - for example, arrows going around corners seem to be quite sharp-angled.  However that's really easy to adjust - see this video for some really quite nice arrows following the route (an earlier prototype which nobody but me liked)... https://www.youtube.com/watch?v=wvE9HUDMqSM

**Comment by Norbert80** [comment]
Sorry. But I don't like it too. The arrows look like they have tails that are more thin at their ends.

**Comment by andrew-davie** [comment]
@Norbert80 you are *completely* missing the point. The point is that the arrows DO closely match the line, pretty much exactly. In any case I'm not pushing my code so you can fix the original problem however you like.

**Comment by Norbert80** [comment]
@vshcherb I started using OsmAnd for bike navigation and I often see the route line vanishing. The zoom level I use most time for biking seems to be the same or nearly the same like in my initial screen shot of this issue.

Victor, could you please raise the priority on this bug?

**Comment by vshcherb** [comment]
Yes the bug is known and Andrew is perfectly right what was going on. We had some good conversations with Andrew but it took us a while to get a previous merge-request in a mergeable state so I switched priorities to OSM Live. Hopefully OSM Live is back in the good state, all the jobs are about to finish and I'm about to get back to 3 pull-requests (altitude/speed display of GPX - Andrew, route simplification - Andrew and bicycle routing). 

It is very sad to hear that it took so long for me to get back. I think the approach, Andrew took, is perfectly correct, though it is a quite big refactoring that's why it requires code review.  

So to summarize, I will try to promise that coming week I will do all my best to get it through. It is in a priority list unless something goes bad in production.
P.S. : about altitude/speed drawing we still need to come up with User Interaction flow though it should not stop us to merge pull-request with code options.

**Comment by Norbert80** [comment]
@vshcherb thanks for your clarification.

@andrew-davie sorry for the misunderstanding. Could you please publish your pull-request again?

**Comment by andrew-davie** [comment]
@Norbert80 sure, no problem. I’ll send the pull request in a couple of days.

**Comment by sonora** [comment]
Is this still open?

**Comment by fkoester** [comment]
@sonora 
> Is this still open?

I did not notice this behavior any more since the last couple of releases.

**Comment by Norbert80** [comment]
During this year's summer I hae seen it often during bike navigation on a cheap Huawei device.

**Comment by sonora** [comment]
I will close it, we can always re-open should it show up again.
